cmd/gitserver/server/server.go

```
@@ -2178,17 +2366,103 @@ func (s *Server) doClone(ctx context.Context, repo api.RepoName, dir common.GitD
 		cmd.Env = os.Environ()
 	}
 
+	pr := progressPipe.reader
+	pw := progressPipe.writer
+
+	go readCloneProgress(s.DB, logger, newURLRedactor(remoteURL), lock, pr, repo)
+
+	// If this repo is meant to be deduplicated, let us first check if the pool repo exists or not.
+	// If it doesn't clone it out of step and then continue with git fetch in this repo which will
+	// be efficient because this won't fetch the objects already in the pool repo.
+	if isDedupeClone && opts.DeduplicatedCloneOptions.dedupeCloneType != dedupePool {
+		// Update or create pool.
+		poolDir := opts.DeduplicatedCloneOptions.poolDir
+
+		_, err := os.Stat(string(poolDir))
+		if err != nil {
+			if !errors.Is(err, os.ErrNotExist) {
+				return errors.Wrap(err, "failed to determine if pool repo exists, deduplicated storage will not be available for this repo")
```

Can wrapping an original error be redundant here? Will this error be surfaced anywhere on the UI or will it just go straight to logs?

---

I don't think this is redundant wrapping even if it goes straight to logs. Logs are meant to be checked by humans after all. It can be useful to know what was the lookup in relation to that failed in stead of a cryptic os.Stat error or something.

---

I don't have a concern about logs. I like the wrapping in logs, TBH. Kinda gives a feel of a stack trace :)

My concern is if this error eventually surfaced on the UI. In this case I don't think that wrapping will look good (like most our gitserver errors 😢)

---

Hmm. Do you have a suggestion on how to frame this error then? From my perspective I wanted the error to be describe:

1. What failed
2. What is the consequence of the error

Not wrapping it would not satisfy both of the points. And hiding the error would make it cumbersome to figure out the root cause.

---

Not wrapping should be fine, IMO. Do you have an example of an error that can be wrapped here? If it is not "exit 128" kind, then maybe we can add it to the error string itself as a: `"Repo deduplication error: failed to determine if pool repo exists. Caused by: %s"`, where %s is the error being wrapped.

WDYT?

---

> Do you have an example of an error that can be wrapped here? I

Nope. It could be anything that the stdlib may encounter. Maybe something like permission denied for example. It's not enough to just show that error without adding more context. 

Creating a new error hides the real error for callers of this function who may be able to unwrap it and decide what to do with it. In Go it's preferred to wrap errors vs sprintf-ing them into new errors.